File: src/main/java/main/sorting/impl/HeapSort.java
package main.sorting.impl;

import main.sorting.Sort;

public class HeapSort implements Sort<int[]> {
	private int[] table;

	@Override
	public int[] getSortedData(int[] tab) {
		this.table = tab;
		sort();
		return table;
	}

	private void sort() {
		for (int i = table.length / 2 - 1; i >= 0; i--) {
			heapify(i, table.length);
		}
		for (int i = table.length - 1; i >= 0; i--) {
			int temp = table[0];
			table[0] = table[i];
			table[i] = temp;
			heapify(0, i);
		}
	}

	private void heapify(int i, int n) {
		int max = i;

		int left = getLeft(i);
		if (left < n && table[left] > table[max]) {
			max = left;
		}

		int right = getRight(i);
		if (right < n && table[right] > table[max]) {
			max = right;
		}

		if (max != i) {
			int temp = table[max];
			table[max] = table[i];
			table[i] = temp;

			heapify(max, n);
		}
	}

	private int getLeft(int i) {
		return 2 * i + 1;
	}

	private int getRight(int i) {
		return 2 * i + 2;
	}
}

File: src/main/java/main/sorting_kotlin/impl/HeapSort.kt
package main.sorting_kotlin.impl

import main.sorting_kotlin.Sort

class HeapSort : Sort<IntArray> {
    override fun getSorted(table: IntArray): IntArray {
        for (i in table.size / 2 - 1 downTo  0) {
            heapify(table, i, table.size)
        }
        for (i in table.size - 1 downTo 0) {
            swap(table, 0, i)
            heapify(table, 0, i)
        }
        return table
    }

    fun heapify(table: IntArray, i: Int, n: Int){
        var max = i

        if (left(i) < n && table[left(i)] > table[max]) {
            max = left(i)
        }
        if (right(i) < n && table[right(i)] > table[max]) {
            max = right(i)
        }
        if (max != i) {
            swap(table, max, i)
            heapify(table, max, n)
        }
    }

    fun left(i: Int): Int {
        return 2 * i + 1
    }

    fun right(i: Int): Int {
        return 2 * i + 2
    }

    fun swap(table: IntArray, a: Int, b: Int) {
        val temp = table[a]
        table[a] = table[b]
        table[b] = temp
    }
}
